### PR TITLE
Ocd 2912 - Create a basic complaint

### PR DESCRIPTION
**Pull request by tmy1313**

### CODE REVIEW

File: PENDING_RELEASE_NOTES.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ _Date TBD_
 
 ### New features
 * Updated user and user action activity reports to use activity metadata
+* Add ability to view/add/edit surveillance complaints for quarterly reports
```
Comment (andlar) [inline PENDING_RELEASE_NOTES.md]: This is behind a flag, right? If so, this line should be in FLAGGED RELEASE NOTES

File: src/app/components/surveillance/complaint/complaint.component.js [hunk 2]
```diff
@@ -2,16 +2,53 @@ export const SurveillanceComplaintComponent = {
     templateUrl: 'chpl.components/surveillance/complaint/complaint.html',
     bindings: {
         complaint: '<',
-        isEditing: '<',
+        complaints: '<',
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.component.js]: This component is a dumb component for a single complaint. It should have two basic modes: View, and Edit. View might have a sub-mode for "view summary", and edit might have a separate mode if creating from scratch, but this component shouldn't be handling a list of complaints at all

File: src/app/components/surveillance/complaint/complaint.html [hunk 3]
```diff
@@ -1,3 +1,141 @@
 <div id="chpl-complaint" ng-form="$ctrl.form">
-  Complaint "dumb component"
+  <div ng-if="$ctrl.mode === $ctrl.modes.SELECT">
+    <table class="table table-condensed">
+      <thead>
+        <tr>
+          <th scope="col">ONC-ACB</th>
+          <th scope="col">Type</th>
+          <th scope="col">Status</th>
+          <th scope="col">ONC Complaint Id</th>
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]:
```suggestion
          <th scope="col">ONC Complaint ID</th>
```

There are about seven "Id"s in our codebase that I can find, and dozens of "ID"s. Would like to try to keep consistent

File: src/app/components/surveillance/complaint/complaint.html [hunk 4]
```diff
@@ -1,3 +1,141 @@
 <div id="chpl-complaint" ng-form="$ctrl.form">
-  Complaint "dumb component"
+  <div ng-if="$ctrl.mode === $ctrl.modes.SELECT">
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: This is the table that should be in the "smart" component, under the pages directory

File: src/app/components/surveillance/complaint/complaint.html [hunk 5]
```diff
@@ -1,3 +1,141 @@
...
+          <td>
+            <button class="btn btn-link" ng-click="$ctrl.selectComplaint(complaint)"><i class="fa fa-edit"></i><span class="sr-only">Edit complaint</span></button>
+          </td>
+        </tr>
+      </tbody>
+    </table>
+  </div>
+  <div ng-if="$ctrl.mode === $ctrl.modes.EDIT || $ctrl.mode === $ctrl.modes.ADD">
+    <div class="form-group" ng-if="$ctrl.hasAnyRole(['ROLE_ADMIN', 'ROLE_ONC'])">
+      <label class="control-label" for="certificationBody">ONC-ACB<span class="text-danger">*</span></label>
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: If in "EDIT" mode, there won't be a control for this label to connect to. Also, the target of the "for" should be the ID of the control, not the name. The name is used in the `$form` angularJS object for validation stuff

File: src/app/components/surveillance/complaint/complaint.html [hunk 6]
```diff
@@ -1,3 +1,141 @@
...
+      <div ng-if="$ctrl.mode === $ctrl.modes.EDIT" name="certificationBody">
+        {{ $ctrl.complaint.certificationBody.name }}
+      </div>
+      <div ng-if="$ctrl.mode === $ctrl.modes.ADD">
+        <select ng-model="$ctrl.complaint.certificationBody"
+          id="certification-body" name="certificationBody" class="input-sm form-control"
+          ng-options="acb.name for acb in $ctrl.certificationBodies | orderBy: 'name' | filter: {retired: false} track by acb.id"
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: Where do the certificationBodies come from? If it's the general "get all" then that would let an ACB user apply a complaint to an ACB not of their own

File: src/app/components/surveillance/complaint/complaint.html [hunk 7]
```diff
@@ -1,3 +1,141 @@
...
+    <div class="form-group">
+      <label class="control-label" for="acbComplaintId">ONC-ACB Complaint Id</label>
+      <div>
+        <input class="input-sm form-control" type="text"  ng-model="$ctrl.complaint.acbComplaintId" id="acb-complaint-id" name="acbComplaintId" ></input>
+      </div>
+    </div>
+    <div class="form-group">
+      <label class="control-label" for="receivedDate">Received Date</label>
+      <div>
+          <input type="date" class="input-sm form-control" ng-model="$ctrl.complaint.formattedReceivedDate" name="receivedDate" id="received-date" placeholder="yyyy-MM-dd"></input>
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: There are some other places in the codebase where we use a uib-datepicker component to make date choosing easier

File: src/app/components/surveillance/complaint/complaint.html [hunk 8]
```diff
@@ -1,3 +1,141 @@
...
+    </div>
+    <div class="form-group">
+      <label class="checkbox-inline" for="oncAtlContacted">
+        <input type="checkbox" id="onc-atl-contacted" name="oncAtlContacted" ng-model="$ctrl.complaint.oncAtlContacted"> ONC-ATL Contacted?
+      </label>
+    </div>
+    <div class="text-danger text-left" ng-if="$ctrl.form.$invalid && $ctrl.showFormErrors">
+      The following error(s) must be fixed:
+      <ul>
+        <li ng-if="$ctrl.form.certificationBody.$error.required">ONC-ACB is required</li>
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: There were five elements that were required, according to the Teams conversation, but I only see one here

File: src/app/components/surveillance/complaint/complaint.html [hunk 9]
```diff
@@ -1,3 +1,141 @@
...
+        <li ng-repeat="message in $ctrl.errorMessages">{{ message }}</li>
+      </ul>
+    </div>
+    <div class="text-danger text-left" ng-if="$ctrl.errorMessages.length > 0">
+        The following error(s) must be fixed:
+        <ul>
+          <li ng-repeat="message in $ctrl.errorMessages">{{ message }}</li>
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: Will these messages show twice?

File: src/app/components/surveillance/complaint/complaint.component.js [hunk 10]
```diff
@@ -2,16 +2,53 @@ export const SurveillanceComplaintComponent = {
...
+        mode: '<',
+        complaintTypes: '<',
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.component.js]: Each of the bindings that have an '<' should be referenced in the $onChanges function, to make a copy of them and break the connection back to the smart component above

File: src/app/pages/surveillance/complaints/complaints.component.js [hunk 11]
```diff
@@ -2,10 +2,134 @@ export const SurveillanceComplaintsComponent = {
...
+            this.refreshComplaints();
+            this.refreshComplaintTypes();
+            this.refreshComplaintStatusTypes();
+            this.refreshCertificationBodies();
+        }
+
+        deleteComplaint (complaint) {
+            this.clearErrorMessages();
+            this.networkService.deleteComplaint(complaint.id).then(() => {
+                this.complaint = {};
```
Comment (andlar) [inline src/app/pages/surveillance/complaints/complaints.component.js]: I don't think `this` exists in this callback. I think this should be `that`, and you'll need a `let that = this;` line just before the network call

File: src/app/pages/surveillance/complaints/complaints.component.js [hunk 12]
```diff
@@ -2,10 +2,134 @@ export const SurveillanceComplaintsComponent = {
...
+                    if (error.status === 400) {
+                        this.errorMessages = error.data.errorMessages;
+                    }
+                });
+        }
+
+        createComplaint (complaint) {
+            this.clearErrorMessages();
+            // default the status to Open
+            complaint.complaintStatusType = this.getComplaintStatusType('Open');
```
Comment (andlar) [inline src/app/pages/surveillance/complaints/complaints.component.js]: What happens if a user creates a complaint with a status of "Closed"? Will this overwrite them?

File: src/app/pages/surveillance/complaints/complaints.html [hunk 13]
```diff
@@ -1,6 +1,26 @@
...
+      Add New Complaint
+    </button>
+
+    <chpl-surveillance-complaint
+      mode="$ctrl.currentMode"
+      complaints="$ctrl.complaints"
+      complaint="$ctrl.complaint"
+      complaint-types="$ctrl.complaintTypes"
+      complaint-status-types="$ctrl.complaintStatusTypes"
+      certification_bodies="$ctrl.certificationBodies"
```
Comment (andlar) [inline src/app/pages/surveillance/complaints/complaints.html]:
```suggestion
      certification-bodies="$ctrl.certificationBodies"
```

File: src/app/components/surveillance/complaint/complaint.html [hunk 14]
```diff
@@ -1,3 +1,141 @@
...
+        </ul>
+      </div>
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: Indentation got weird here. Could be missing a </div>, or bad indentation somewhere else

File: src/app/components/surveillance/complaint/complaint.html [hunk 15]
```diff
@@ -1,3 +1,141 @@
...
+          <td>{{ complaint.formattedReceivedDate | date : 'MMM d, y' : 'UTC' }}</td>
+          <td>
+            <button class="btn btn-link" ng-click="$ctrl.selectComplaint(complaint)"><i class="fa fa-edit"></i><span class="sr-only">Edit complaint</span></button>
+          </td>
+        </tr>
+      </tbody>
+    </table>
+  </div>
+  <div ng-if="$ctrl.mode === $ctrl.modes.EDIT || $ctrl.mode === $ctrl.modes.ADD">
+    <div class="form-group" ng-if="$ctrl.hasAnyRole(['ROLE_ADMIN', 'ROLE_ONC'])">
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: What happens if a ROLE_ACB user has access to two ACBs?

File: src/app/pages/surveillance/complaints/complaints.component.js (hunk 12 above)
Comment (tmy1313) [reply src/app/pages/surveillance/complaints/complaints.component.js]: User cannot create a complaint with a status of closed based on statements from Heather.

File: src/app/components/surveillance/complaint/complaint.html [hunk 16]
```diff
@@ -1,3 +1,136 @@
 <div id="chpl-complaint" ng-form="$ctrl.form">
-  Complaint "dumb component"
-</div>
+  <h2 ng-if="$ctrl.currentMode === $ctrl.modes.EDIT">Edit Complaint</h2>
+  <h2 ng-if="$ctrl.currentMode === $ctrl.modes.ADD">Add Complaint</h2>
+  <div class="form-group" ng-if="$ctrl.hasAnyRole(['ROLE_ADMIN', 'ROLE_ONC'])">
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: What happens if ROLE_ACB is logged in? Do they see the ACB? What happens if they have access to two ACBs?

File: src/app/pages/surveillance/complaints/complaints.html [hunk 17]
```diff
@@ -1,6 +1,49 @@
 <div class="container-fluid">
-  <div class="row main-content" id="main-content" tabindex="-1">
-    Complaints "smart component"
-    <chpl-surveillance-complaint></chpl-surveillance-complaint>
+  <div class="main-content" id="main-content" tabindex="-1">
```
Comment (andlar) [inline src/app/pages/surveillance/complaints/complaints.html]: The bootstrap model is to have a container (line 1), then rows with 12-column grids. We've mostly just gotten into col-md-12 columns, but should still have something there

File: src/app/components/surveillance/complaint/complaint.html [hunk 18]
```diff
@@ -1,3 +1,142 @@
...
+    <label class="control-label" for="complaintStatusType">Complaint Status Type</label>
+    <div>
+      <select ng-model="$ctrl.complaint.complaintStatusType" id="complaint-status-type" name="complaintStatusType"
+        class="input-sm form-control"
+        ng-options="cst.name for cst in $ctrl.complaintStatusTypes | orderBy: 'name' track by cst.id">
+      </select>
+    </div>
+  </div>
+  <div class="form-group">
+    <label class="control-label" for="oncComplaintId">ONC Complaint ID</label>
```
Comment (andlar) [inline src/app/components/surveillance/complaint/complaint.html]: A `<label>`'s "for" attribute is supposed to target the id of the element that the label is for, not the name

**Pull request merged**